include/swift/AST/Attr.h

```diff
@@ -182,60 +183,82 @@ enum class DeclKind : uint8_t;
 class DeclAttribute : public AttributeBase {
   friend class DeclAttributes;
 protected:
-  class DeclAttrBitFields {
-    friend class DeclAttribute;
-
-    // The kind.
-    unsigned Kind : 8;
-
-    // Whether this attribute was implicitly added.
-    unsigned Implicit : 1;
-
-    unsigned Invalid : 1;
-  };
-  enum { NumDeclAttrBits = 10 };
-  static_assert(NumDeclAttrBits <= 32, "fits in an unsigned");
-
-  class ObjCAttrBitFields {
-    friend class ObjCAttr;
-    unsigned : NumDeclAttrBits;
-
-    /// Whether this attribute has location information that trails the main
-    /// record, which contains the locations of the parentheses and any names.
-    unsigned HasTrailingLocationInfo : 1;
-
-    /// Whether the name is implicit, produced as the result of caching.
-    unsigned ImplicitName : 1;
-
-    /// Whether the @objc was inferred using Swift 3's deprecated inference
-    /// rules.
-    unsigned Swift3Inferred : 1;
-  };
-  enum { NumObjCAttrBits = NumDeclAttrBits + 3 };
-  static_assert(NumObjCAttrBits <= 32, "fits in an unsigned");
-
-  class AccessControlAttrBitFields {
-    friend class AbstractAccessControlAttr;
-    unsigned : NumDeclAttrBits;
-
-    unsigned AccessLevel : 3;
-  };
-  enum { NumAccessControlAttrBits = NumDeclAttrBits + 3 };
-  static_assert(NumAccessControlAttrBits <= 32, "fits in an unsigned");
-
   union {
-    DeclAttrBitFields DeclAttrBits;
-    ObjCAttrBitFields ObjCAttrBits;
-    AccessControlAttrBitFields AccessControlAttrBits;
-  };
+    uint64_t OpaqueBits;
+
+    SWIFT_INLINE_BITFIELD_BASE(DeclAttribute, bitmax(NumDeclAttrKindBits,8)+1+1,
+      Kind : bitmax(NumDeclAttrKindBits,8),
+      // Whether this attribute was implicitly added.
+      Implicit : 1,
+
+      Invalid : 1
+    );
+
+    SWIFT_INLINE_BITFIELD(ObjCAttr, DeclAttribute, 1+1+1,
+      /// Whether this attribute has location information that trails the main
+      /// record, which contains the locations of the parentheses and any names.
+      HasTrailingLocationInfo : 1,
+
+      /// Whether the name is implicit, produced as the result of caching.
+      ImplicitName : 1,
+
+      /// Whether the @objc was inferred using Swift 3's deprecated inference
+      /// rules.
+      Swift3Inferred : 1
+    );
+
+    SWIFT_INLINE_BITFIELD(AbstractAccessControlAttr, DeclAttribute, 3,
+      AccessLevel : 3
+    );
+
+    SWIFT_INLINE_BITFIELD_FULL(AlignmentAttr, DeclAttribute, 32,
+      : NumPadBits,
+      // The alignment value.
+      Value : 32
```

Comment:
This (and the other 32-bit value below) increases the size of the bitfield from 32 to 64. Is that a win? (Are we getting that anyway because of alignment?)

Reply:
This came up when the bitfield macros were introduced. Pointers force 64-bit alignment and pointers are difficult to avoid in AST nodes. In this specific case, the superclass `AttributeBase` contains pointers, and `DeclAttribute` itself contains a "next" pointer *after* the bitfield, therefore (and once again), a 64-bit inline bitfield is "free".

Reply:
Thanks for the (re)-explanation!